File: artifacts/avl-tree/refactor/avl_example.c
#include <avl_bf.h>
#include <avl_data.h>
#include <avl_example.c>
#include <stdio.h>
#include <stdlib.h>

enum avlbf
{
  LEFTHEAVY = -1,
  BALANCED = 0,
  RIGHTHEAVY = 1
};
enum avltraversal
{
  PREORDER,
  INORDER,
  POSTORDER
};
typedef struct avlnode
{
  struct avlnode *left;
  struct avlnode *right;
  struct avlnode *parent;
  char bf;
  void *data;
} avlnode;
typedef struct 
{
  int (*compare)(const void *, const void *);
  void (*print)(void *);
  void (*destroy)(void *);
  avlnode root;
  avlnode nil;
  avlnode *min;
} avltree;
avltree *avl_create(int (*compare_func)(const void *, const void *), void (*destroy_func)(void *));
void avl_destroy(avltree *avlt);
avlnode *avl_find(avltree *avlt, void *data);
avlnode *avl_successor(avltree *avlt, avlnode *node);
int avl_apply(avltree *avlt, avlnode *node, int (*func)(void *, void *), void *cookie, enum avltraversal order);
void avl_print(avltree *avlt, void (*print_func)(void *));
avlnode *avl_insert(avltree *avlt, void *data);
void *avl_delete(avltree *avlt, avlnode *node, int keep);
int avl_check_order(avltree *avlt, void *min, void *max);
int avl_check_height(avltree *avlt);
typedef struct 
{
  int key;
} mydata;
mydata *makedata(int key);
int compare_func(const void *d1, const void *d2);
void destroy_func(void *d);
void print_func(void *d);
void print_char_func(void *d);
int main(int argc, char *argv[])
{
  avltree *avlt;
  unsigned int avlt_idx = 0;
  if ((avlt_idx = avl_create(compare_func, destroy_func)) == 0)
  {
    fprintf(stderr, "create AVL tree failed\n");
    return 1;
  }
  char a[] = {'R', 'E', 'D', 'S', 'O', 'X', 'C', 'U', 'B', 'T'};
  int i;
  mydata *data;
  unsigned int data_idx = 0;
  for (i = 0; i < ((sizeof(a)) / (sizeof(a[0]))); i += 1)
  {
    if (((data_idx = makedata(a[i])) == 0) || (avl_insert(avlt, data) == 0))
    {
      fprintf(stderr, "insert %c: out of memory\n", a[i]);
      free(data);
      break;
    }
    printf("insert %c", a[i]);
    avl_print(avlt, print_char_func);
    printf("\n");
  }

  avlnode *node;
  unsigned int node_idx = 0;
  mydata query;
  query.key = 'O';
  printf("delete %c", query.key);
  if ((node_idx = avl_find(avlt, &query)) != 0)
  {
    avl_delete(avlt, node, 0);
  }
  avl_print(avlt, print_char_func);
  while (node_idx = avlt->min)
  {
    printf("\ndelete ");
    print_char_func(node->data);
    avl_delete(avlt, node, 0);
    avl_print(avlt, print_char_func);
  }

  avl_destroy(avlt);
  return 0;
}


enum avlbf
{
  LEFTHEAVY = -1,
  BALANCED = 0,
  RIGHTHEAVY = 1
};
enum avltraversal
{
  PREORDER,
  INORDER,
  POSTORDER
};
typedef struct avlnode
{
  struct avlnode *left;
  struct avlnode *right;
  struct avlnode *parent;
  char bf;
  void *data;
} avlnode;
typedef struct 
{
  int (*compare)(const void *, const void *);
  void (*print)(void *);
  void (*destroy)(void *);
  avlnode root;
  avlnode nil;
  avlnode *min;
} avltree;
avltree *avl_create(int (*compare_func)(const void *, const void *), void (*destroy_func)(void *));
void avl_destroy(avltree *avlt);
avlnode *avl_find(avltree *avlt, void *data);
avlnode *avl_successor(avltree *avlt, avlnode *node);
int avl_apply(avltree *avlt, avlnode *node, int (*func)(void *, void *), void *cookie, enum avltraversal order);
void avl_print(avltree *avlt, void (*print_func)(void *));
avlnode *avl_insert(avltree *avlt, void *data);
void *avl_delete(avltree *avlt, avlnode *node, int keep);
int avl_check_order(avltree *avlt, void *min, void *max);
int avl_check_height(avltree *avlt);
typedef struct 
{
  int key;
} mydata;
mydata *makedata(int key);
int compare_func(const void *d1, const void *d2);
void destroy_func(void *d);
void print_func(void *d);
void print_char_func(void *d);
int main(int argc, char *argv[])
{
  avltree *avlt;
  unsigned int avlt_idx = 0;
  if ((avlt_idx = avl_create(compare_func, destroy_func)) == 0)
  {
    fprintf(stderr, "create AVL tree failed\n");
    return 1;
  }
  char a[] = {'R', 'E', 'D', 'S', 'O', 'X', 'C', 'U', 'B', 'T'};
  int i;
  mydata *data;
  unsigned int data_idx = 0;
  for (i = 0; i < ((sizeof(a)) / (sizeof(a[0]))); i += 1)
  {
    if (((data_idx = makedata(a[i])) == 0) || (avl_insert(avlt, data) == 0))
    {
      fprintf(stderr, "insert %c: out of memory\n", a[i]);
      free(data);
      break;
    }
    printf("insert %c", a[i]);
    avl_print(avlt, print_char_func);
    printf("\n");
  }

  avlnode *node;
  unsigned int node_idx = 0;
  mydata query;
  query.key = 'O';
  printf("delete %c", query.key);
  if ((node_idx = avl_find(avlt, &query)) != 0)
  {
    avl_delete(avlt, node, 0);
  }
  avl_print(avlt, print_char_func);
  while (node_idx = avlt->min)
  {
    printf("\ndelete ");
    print_char_func(node->data);
    avl_delete(avlt, node, 0);
    avl_print(avlt, print_char_func);
  }

  avl_destroy(avlt);
  return 0;
}

File: artifacts/avl-tree/translate/src/avl_example.rs
use translate::*;
use std::env;
use std::any::Any;
pub fn main() {
    let avlt: Option<Box<AvlTree>> = avl_create(
        Some(Box::new(|a: &dyn Any, b: &dyn Any| -> i32 {
            let a = a.downcast_ref::<Mydata>().unwrap();
            let b = b.downcast_ref::<Mydata>().unwrap();
            compare_func(Some(a), Some(b))
        })),
        Some(Box::new(|a: &dyn Any| {
            let data = Box::new(a.downcast_ref::<Mydata>().unwrap().clone());
            destroy_func(Some(Box::new(*data)))
        })),
    );

    if avlt.is_none() {
        eprintln!("create AVL tree failed");
        return;
    }

    let mut avlt = avlt.unwrap();
    let a = ['R', 'E', 'D', 'S', 'O', 'X', 'C', 'U', 'B', 'T'];
    
    for &c in &a {
        let data = match makedata(c as i32) {
            Some(d) => Box::new(d) as Box<dyn Any>,
            None => {
                eprintln!("insert {}: out of memory", c as char);
                break;
            }
        };

        if avl_insert(&mut avlt, data).is_none() {
            eprintln!("insert {}: out of memory", c as char);
            break;
        }

        println!("insert {}", c as char);
        avl_print(&avlt, &|opt: Option<&dyn Any>| {
            if let Some(d) = opt {
                let d = d.downcast_ref::<Mydata>().unwrap();
                print_char_func(Some(d));
            }
        });
        println!();
    }

    let query = Mydata { key: 'O' as i32 };
    println!("delete {}", query.key as u8 as char);
    
    // Fix: Store the node data before deletion
    if let Some(node) = avl_find(&avlt, &query) {
        let node_data = node.data.as_ref().unwrap().downcast_ref::<Mydata>().unwrap().clone();
        let mut node_clone = AvlNode {
            left: node.left.as_ref().map(|_| panic!("Cannot clone Box<AvlNode>")),
            right: node.right.as_ref().map(|_| panic!("Cannot clone Box<AvlNode>")),
            parent: node.parent.as_ref().map(|_| panic!("Cannot clone Box<AvlNode>")),
            bf: node.bf,
            data: Some(Box::new(node_data)),
        };
        avl_delete(&mut avlt, &mut node_clone, 0);
    }
    
    avl_print(&avlt, &|opt: Option<&dyn Any>| {
        if let Some(d) = opt {
            let d = d.downcast_ref::<Mydata>().unwrap();
            print_char_func(Some(d));
        }
    });
    
    // Fix: Properly handle the min node deletion
    while let Some(node) = avlt.min.take() {
        print!("\ndelete ");
        let node_data = node.data.as_ref().unwrap().downcast_ref::<Mydata>().unwrap().clone();
        print_char_func(Some(&node_data));
        let mut node_clone = AvlNode {
            left: None,
            right: None,
            parent: None,
            bf: node.bf,
            data: Some(Box::new(node_data)),
        };
        avl_delete(&mut avlt, &mut node_clone, 0);
        avl_print(&avlt, &|opt: Option<&dyn Any>| {
            if let Some(d) = opt {
                let d = d.downcast_ref::<Mydata>().unwrap();
                print_char_func(Some(d));
            }
        });
    }

    avl_destroy(avlt);
}
